xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +135,793 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+}
+
+func (b *outlierDetectionBalancer) Target() string {
+	return b.cc.Target()
+}
+
+func max(x, y int64) int64 {
+	if x < y {
+		return y
+	}
+	return x
+}
+
+func min(x, y int64) int64 {
+	if x < y {
+		return x
+	}
+	return y
+}
+
+// handleSubConnUpdate stores the recent state and forward the update
+// if the SubConn is not ejected.
+func (b *outlierDetectionBalancer) handleSubConnUpdate(u *scUpdate) {
+	scw := u.scw
+	scw.latestState = u.state
+	b.childMu.Lock()
+	if !scw.ejected && b.child != nil {
+		b.child.UpdateSubConnState(scw, u.state)
+	}
+	b.childMu.Unlock()
+}
+
+// handleEjectedUpdate handles any SubConns that get ejected/unejected, and
+// forwards the appropriate corresponding subConnState to the child policy.
+func (b *outlierDetectionBalancer) handleEjectedUpdate(u *ejectedUpdate) {
+	scw := u.scw
+	scw.ejected = u.ejected
+	var stateToUpdate balancer.SubConnState
+	if u.ejected {
+		// "The wrapper will report a state update with the
+		// TRANSIENT_FAILURE state, and will stop passing along
+		// updates from the underlying subchannel."
+		stateToUpdate = balancer.SubConnState{
+			ConnectivityState: connectivity.TransientFailure,
+		}
+	} else {
+		// "The wrapper will report a state update with the latest
+		// update from the underlying subchannel, and resume passing
+		// along updates from the underlying subchannel."
+		stateToUpdate = scw.latestState // If this has never been written to will send connectivity IDLE which seems fine to me
+	}
+	b.childMu.Lock()
+	if b.child != nil {
+		b.child.UpdateSubConnState(scw, stateToUpdate)
+	}
+	b.childMu.Unlock()
+}
+
+// handleChildStateUpdate forwards the picker update wrapped in a wrapped picker
+// with the noop picker bit present.
+func (b *outlierDetectionBalancer) handleChildStateUpdate(u balancer.State) {
+	b.childState = u
+	b.mu.Lock()
+	noopCfg := b.noopConfig()
+	b.mu.Unlock()
+	b.recentPickerNoop = noopCfg
+	b.cc.UpdateState(balancer.State{
+		ConnectivityState: b.childState.ConnectivityState,
+		// The outlier_detection LB policy will provide a picker that delegates to
+		// the child policy's picker, and when the request finishes, increment the
+		// corresponding counter in the map entry referenced by the subchannel
+		// wrapper that was picked.
+		Picker: &wrappedPicker{
+			childPicker: b.childState.Picker,
+			// If both the `success_rate_ejection` and
+			// `failure_percentage_ejection` fields are unset in the
+			// configuration, the picker should not do that counting.
+			noopPicker: noopCfg,
+		},
+	})
+}
+
+// handleLBConfigUpdate compares whether the new config is a noop config or not,
+// to the noop bit in the picker if present. It updates the picker if this bit
+// changed compared to the picker currently in use.
+func (b *outlierDetectionBalancer) handleLBConfigUpdate(u *LBConfig) {
+	noopCfg := u.SuccessRateEjection == nil && u.FailurePercentageEjection == nil
+	if b.childState.Picker != nil && noopCfg != b.recentPickerNoop {
```

Comment:
Why is the picker updated only if these conditions are met? Does this mean that the outlier_detection LB policy does not report a picker to the channel until the child reports one? If so, why?